designs/0034-jacobian.md

```diff
@@ -0,0 +1,381 @@
+   ReturnType constrain(UnconstrainType, OtherArgs...) {
+     // ... transformation logic ...
+     // Increment Jacobian
+     jacobian += ...;
+     return ...;
+   }
+
+   // Transform from constrained to unconstrained space
+   ReturnType unconstrain(UnconstrainType, OtherArgs...) {
+     // ... inverse transformation logic ...
+     return ...;
+   }
+
+   // Validate data
+   int validate(UnconstrainType, OtherArgs...) {
+     // ... validation logic ...
+     return ...; // typically returns a boolean value
+   }
+ }
+}
+```
+
+### Using the jacobian Keyword
+
+The `jacobian` keyword is introduced within the `constraint` block to allow the Jacobian accumulator to be incremented by the log absolute determinant of the Jacobian of the constraining transform. 
+This keyword behaves similarly to the `target` keyword, allowing users to account for the change of variables in the probability density function. 
+The `jacobian` keyword is available in both the `constraints` block and the `transformed parameters` block.
+
+### Declaring Variables with User-defined Constraints
+
+Variables can be declared with user-defined constraints in the data, transformed data, parameters, transformed parameters, and generated quantities blocks. 
+The syntax for declaring a variable with a user-defined constraint is as follows:
+
+```stan
+type<constraint constraint_name(args...)> variable_name;
+```
+
+Here, `constraint` denotes a user-defined constraint and `constraint_name(args...)` specifies the name of the constraint along with any arguments it requires.
+
+### Example Usage
+
+```stan
+constraints {
+ upper_bound {
+   real constrain(real x, real upper_bound) {
+     jacobian += x;
+     return upper_bound - exp(x);
+   }
+   real unconstrain(real x, real upper_bound) {
+     return log(upper_bound - x);
+   }
+   vector constrain(vector x, real upper_bound) {
+     jacobian += sum(x);
+     return upper_bound - exp(x);
+   }
+   vector unconstrain(vector x, real upper_bound) {
+     return log(upper_bound - x);
+   }
+   int validate(real x, real upper_bound) {
+     return x < upper_bound;
+   }
+ }
+}
+data {
+ real a;
+ real u_bound;
+ int N;
+}
+
+
+transformed data {
+ // Data will use the validate() function from the user-defined constraint
+ real<constraint upper_bound(u_bound)> c = a;
+}
+
+
+parameters {
+ // User-defined constraint applied to parameters alpha and b
+ real<constraint upper_bound(u_bound)> alpha;
+ vector<constraint upper_bound(u_bound)>[N] b;
+ // In these cases this is the same as calling
+ // real<upper=u_bound> alpha;
+ // vector<upper=u_bound>[N] b;
+}
+```
+
+In this example, `upper_bound` is a user-defined constraint that is applied to the data and parameters.
+The `validate()` function from `upper_bound` is used to validate `a_validated` in the `transformed data` block, and the transformations defined in `upper_bound` are applied to `alpha` and `b` in the parameters block.
```

Comment:
This example highlights the overloading behavior of these functions, but I think it is worth stating more explicitly what it is